daemon/appinfo.c

```diff
@@ -627,51 +646,131 @@ appinfo_clear_permissions(appinfo_t *self)
  * APPINFO_PARSE
  * ------------------------------------------------------------------------- */
 
-bool
-appinfo_parse_desktop(appinfo_t *self)
-{
-    GKeyFile  *ini         = NULL;
-    gchar     *path        = NULL;
-    GError    *err         = NULL;
-    gchar    **permissions = NULL;
+static appinfo_file_t
+appinfo_combined_file_state(appinfo_file_t state1, appinfo_file_t state2)
+{
+    /*  state[12] | UNCHANGED |  CHANGED  |  INVALID  |  DELETED  |  MISSING
+     * ===========#===========#===========#===========#===========#==========
+     *  UNCHANGED | UNCHANGED |  CHANGED  |  INVALID  |  CHANGED  | UNCHANGED
+     * ===========#-----------+-----------+-----------+-----------+-----------
+     *   CHANGED  |  CHANGED  |  CHANGED  |  INVALID  |  CHANGED  |  CHANGED
+     * ===========#-----------+-----------+-----------+-----------+-----------
+     *   INVALID  |  INVALID  |  INVALID  |  INVALID  |  INVALID  |  INVALID
+     * ===========#-----------+-----------+-----------+-----------+-----------
+     *   DELETED  |  CHANGED  |  CHANGED  |  INVALID  |  DELETED  |  DELETED
+     * ===========#-----------+-----------+-----------+-----------+-----------
+     *   MISSING  | UNCHANGED |  CHANGED  |  INVALID  |  DELETED  |  MISSING
+     */
 
-    path = path_from_desktop_name(appinfo_id(self));
+    switch (state1) {
+    case APPINFO_FILE_UNCHANGED:
+        if( state2 == APPINFO_FILE_DELETED )
+            return APPINFO_FILE_CHANGED;
+        if( state2 == APPINFO_FILE_MISSING )
+            return APPINFO_FILE_UNCHANGED;
+        return state2;
+    case APPINFO_FILE_CHANGED:
+        if( state2 == APPINFO_FILE_INVALID )
+            return APPINFO_FILE_INVALID;
+        return APPINFO_FILE_CHANGED;
+    case APPINFO_FILE_INVALID:
+        return APPINFO_FILE_INVALID;
+    case APPINFO_FILE_DELETED:
+        if( state2 <= APPINFO_FILE_CHANGED )
+            return APPINFO_FILE_CHANGED;
+        if( state2 == APPINFO_FILE_INVALID )
+            return APPINFO_FILE_INVALID;
+        return APPINFO_FILE_DELETED;
+    case APPINFO_FILE_MISSING:
+        return state2;
+    }
+    abort();
```

Comment:
gcc seems to complain that "control reaches end of non-void function" without this. AFAICT this can't be reached unless there is a mistake elsewhere. Is there a better option than `abort()`?

Reply:
One possibility would be to add a `default` with fallthrough to one of your `case` clauses. That way gcc should recognise that there's no way to leave the switch and will stop complaining.

Reply:
This can be reached if e.g. state1 is outside of enum's values. But I think abort() could be used here. Or we should return invalid in this case.

Reply:
That state1 outside enum's values is that mistake elsewhere situation. Too bad that compiler can't tell if there is such a mistake there. `default` is not good because then compiler can't check that all branches were explicitly included in `switch`. I think I'll just log an error and return APPINFO_FILE_INVALID so that processing stops there.